### DISCUSSION TITLE
iOS 17.5 (beta) slight change in behaviour when opening AR QuickLook

### DISCUSSION THREAD

**Discussion by milesgreen** [google/model-viewer#4740] [General]
With the new iOS 17.5 beta release (21F5048f) there is a change in behaviour when opening USDZ models to AR QuickLook from Safari.

In Safari 17.5 beta, when a USDZ file is opened to AR QuickLook, the user is presented with a permissions prompt, asking if the user wishes to **Open this 3D model?**

The user can **Cancel** or **Allow**. 

Clicking **Allow** will continue to AR QuickLook as before. 

It's a small change. But potentially interrupts the user experience. This isn't like a site-level permission setting. The user will be asked for every single view of the same - or any other - 3D models on that site and all other sites. 

I haven't seen confirmation if this is the intended new behaviour moving forward. 

So I logged it as a bug anyway to see if there's any more information:
https://bugs.webkit.org/show_bug.cgi?id=272326

![arql-prompt](https://github.com/google/model-viewer/assets/53870482/e5180b2d-e677-4529-9750-d2befbd5f006)

**Comment by milesgreen** [comment]
Feel free to pile on to the bug ticket (https://bugs.webkit.org/show_bug.cgi?id=272326) to voice your opinion if you too don't much like this change...

**Comment by milesgreen** [comment]
Apple have confirmed this behaviour is indeed intentional.

I added my thoughts to the ticket:
https://bugs.webkit.org/show_bug.cgi?id=272326

**Comment by kelvinleusink** [comment]
Hi @milesgreen,

I also noticed that when using activateAR() in safari there is no allow pop-up, it will now ask if you want to view or download the usdz.
We use this for modelviewer viewed on desktop and generate a qr-code so users can scan this with their phone and will be directly pushed to the AR version of the model (quick-look and sceneviewer).

Is there any work around or should Apple fix this?
There is no problem when using chrome on iOS.

![IMG_5779](https://github.com/google/model-viewer/assets/70634810/7d617563-4751-4396-9bdc-4e997a3788ba)
![IMG_5780](https://github.com/google/model-viewer/assets/70634810/568dfdde-92f8-4f6c-9018-a12a630c77c2)
![IMG_5781](https://github.com/google/model-viewer/assets/70634810/3b4fd571-a101-43fe-9481-555d952cccde)

**Reply by milesgreen** [reply]
@kelvinleusink - What you're seeing is a bug in Safari iOS 17.5 (beta).

You're using auto USDZ generation, right? You have configured ModelViewer with a `src` but not an `ios-src`?

And that has become a problem with iOS 17.5 beta.

The bug has been reported and hopefully being worked on. 

The only other solution is to not use auto USDZ generation and to link to an actual USDZ file in the `ios-src`. But sometimes that's not possible. So, hopefully this'll get fixed before public release. 

There's more discussion on this specific bug over here: 
https://github.com/google/model-viewer/issues/4746

**Reply by kelvinleusink** [reply]
Thanks for your reply @milesgreen!
We don't use auto USDZ generation, we are providing a ios-src.

When we use the ar button directly on iOS in modelviewer it works fine, but when we scan the QR code with an iPhone it redirects to an url which activates activarAR() so it opens the usdz directly, which worked super nice before iOS 17.5.

**Reply by kelvinleusink** [reply]
Here you'll find an example: https://cdn-ar.com/nova/viewer/?model=salontafel-luna_hout

**Reply by elalish** [reply]
I think they intentionally disabled the "open AR automatically" thing. That's been blocked by WebXR from the beginning as a security thing. But the part where it fails even when there *is* a user activation is the bug.

**Reply by milesgreen** [reply]
I see. @kelvinleusink, I missed the part about you automatically activating AR on page load. So there's no user interaction. You're just automatically calling activateAR().

Yes, that used to work fine prior to iOS 17.5.

But in iOS 17.5 it throws this weird Download prompt.

I can kind of understand them closing down automatic opening from a security point of view. Even though it did make a nice seamless user journey, to go from scanning QR code to viewing model in AR. 

Still, I would expect the new behaviour to be show the new "Open 3D model" prompt, and give the user the option to Allow / Cancel. Or, just block it altogether. 

But in this case, it's showing this broken Download prompt. 

Locking down auto open seems to be intentional. The error handling, maybe not so much.

**Reply by milesgreen** [reply]
@kelvinleusink - the issue you were seeing when auto opening a USDZ on page load is now fixed with iOS 17.5 beta 4. 

You will still see an "Open 3D model" prompt. I don't think that's going away. But it does then open to ARQL.

**Comment by kelvinleusink** [comment]
@milesgreen Happy to see that the weird download prompt is gone now, thanks for the update!
